add static resource controller

sjenning opened the pull request:
Begins implementing https://github.com/openshift/enhancements/pull/260

This controller deploys the static and templated resources required for the AWS pod identity webhook.

The image pull spec is templated into the `Deployment` using the envvar injected into the operator by the CVO and the `service-ca` is templated into the `MutatingWebhookConfiguration`.

The controller reconciles the assets if they are changed and updates the `MutatingWebhookConfiguration` and does a forced rollout of the `Deployment` if the `service-ca` rotates.

/cc @dgoodwin @joelddiaz @derekwaynecarr @marun @deads2k

sjenning commented:
Still have some RBAC work to do.
/hold

sjenning commented:
/retest

marun commented:
@sjenning Is the `service-ca` related to the service-ca-operator or something else?

dgoodwin commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+
+var (
+	staticFiles = []string{
+		"v4.1.0/aws-pod-identity-webhook/sa.yaml",
```
Is this v4.1.0 the version of aws-pod-identity-webhook? Seems like the directory structure should be inverted, aws-pod-identity-webhook/v4.1.0/sa.yaml

sjenning replied:
yes

sjenning replied:
actually no.  that is referring to the OCP version.  the idea is that if some point in the future, you need access to both the old asset and the new in order to do some non-trivial reconciliation, you'd have both versions.  not sure if i buy the organization/usefulness of that though.  it's just how the kube-* operators do it. eg https://github.com/openshift/cluster-kube-apiserver-operator/tree/master/bindata/v4.1.0/kube-apiserver

dgoodwin commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+		return true
+	}
+	// we also care about service-ca rotation
+	if meta.GetName() == "service-ca" && meta.GetNamespace() == "openshift-config-managed" {
```
Where is the watch setup for this? I can't spot it.

sjenning replied:
that's because it isn't there.  good catch!

sjenning replied:
https://github.com/openshift/cloud-credential-operator/pull/185#discussion_r414758727

dgoodwin commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+	return nil
+}
+
+func isManaged(meta metav1.Object) bool {
```
I assume if we went ahead with a separate config CRD, we could use owner references to more reliably act on what we watch? But as it stands, there is no resource we could mark all of these as owned by.

sjenning replied:
correct

sjenning replied:
https://github.com/openshift/cloud-credential-operator/pull/185#discussion_r414758727

dgoodwin commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+			return err
+		}
+
+		err = c.Watch(&source.Informer{Informer: informer}, &handler.EnqueueRequestForObject{}, p)
```
This is a neat approach to watching everything we created, we may use this in Hive where our operator currently does not watch everything.

Any concerns about memory now that we're watching and thus caching a number of new types cluster wide?

sjenning replied:
I do have concerns about memory and about privilege.  We have to put these abilities on a `ClusterRole` rather than a `Role`, which is not ideal.

I've been looking for a way to limit the scope of the watch to be namespace local, but haven't found it yet.

dgoodwin replied:
@staebler sorry to pull you in here but where did we land with per-namespace watches and limiting cache, if memory serves there was no way to do this?

sjenning replied:
So the cache is at the Manager level, not on the Controller. As other parts of the operator do watch things on a cluster wide basis (namespace creation, secrets, etc), we can't scope down the manager cache.

We could create a new controller-local cache, separate from the Manager cache, and base the Watches on it.  I'll play around with it.

dgoodwin commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+package staticresource
```
Could I request a controller rename, this feels awspodidentity specific and not as generic as staticresource sounds. If we had additional static resource setup to do, we'd probably want that in another controller rather than forcing into this one.

sjenning replied:
Yeah, I thought in the beginning I would make this more generic but... it didn't happen.  I can rename.

dgoodwin commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+var _ reconcile.Reconciler = &staticResourceReconciler{}
+
+func (r *staticResourceReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.WithField("controller", controllerName).Debugf("reconciling after watch event %#v", request)
```
Can use r.logger which already carries the "controller" field context. Several uses of this below.

sjenning replied:
wonderful!

dgoodwin commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+		log.WithField("controller", controllerName).Infof("%s configmap change detected", request.NamespacedName)
+		r.forceRollout = true
+	}
+	r.ReconcileResources()
```
Looks like errors are swallowed here, and thus would not re-try if there was a temporary conflict or similar.

sjenning replied:
yep, needs fixing

dgoodwin commented on manifests/03-deployment.yaml:
```diff
@@ -0,0 +1,75 @@
+        - name: RELEASE_VERSION
+          value: 0.0.1-snapshot
+        - name: AWS_POD_IDENTITY_WEBHOOK_IMAGE
+          value: quay.io/openshift/aws-pod-identity-webhook:latest
```
Is this the value that gets used? Is it acceptable for us to have this kind of implied image which is not reflected in the release image itself? Is the use of latest permanent as well?

sjenning replied:
`quay.io/openshift/aws-pod-identity-webhook:latest` is a placeholder that is replaced by the CVO with the image pull spec during release image creation
https://github.com/openshift/cloud-credential-operator/blob/master/manifests/image-references#L9-L12

dgoodwin commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+	requestedDeployment := resourceread.ReadDeploymentV1OrDie(v410_00_assets.MustAsset("v4.1.0/aws-pod-identity-webhook/deployment.yaml"))
+	requestedDeployment.Spec.Template.Spec.Containers[0].Image = os.Getenv("AWS_POD_IDENTITY_WEBHOOK_IMAGE")
+	resultDeployment, modified, err := resourceapply.ApplyDeployment(r.clientset.AppsV1(), r.eventRecorder, requestedDeployment, r.deploymentGeneration, r.forceRollout)
+	r.forceRollout = false
```
Could some strange things happen here with this forceRollout state being set on the reconciler if we were reconciling concurrently?

sjenning replied:
I think I'm going to pull the watching of the `service-ca` and redeployment logic for this PR.  We will need to add that but for simplicity of review, I think it will be better to leave that out for now, especially since I've switched the cache to be namespace-local (I'll have to add the Manager provided cache back in to access the `service-ca`)

joelddiaz commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+		return err
+	}
+	platformType := platform.GetType(infraStatus)
+	if platformType != configv1.AWSPlatformType {
```
Is it accurate to say that in one potential future (not implying that we'd want to support this...), we could have a cluster running on GCP accessing resources in AWS?

sjenning replied:
@dgoodwin made the same point about this code not being generic.  I'm renaming the package to `awspodidentity`. 

And no, I don't see this code being used on any other platform since it is really only useful for assuming IAM roles, which... I guess you could do from instances on another cloud but... I can't see us supporting that.

joelddiaz commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+	for _, result := range applyResults {
+		if result.Error != nil {
+			log.WithField("controller", controllerName).WithError(result.Error).Errorf("error reconciling %s", result.Type)
+			continue
```
Should we error rather than continue if we've failed to apply some required assets?

joelddiaz commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+
+func (r *staticResourceReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.WithField("controller", controllerName).Debugf("reconciling after watch event %#v", request)
+	if request.Name == "service-ca" && request.Namespace == "openshift-config-managed" {
```
can we constify these names since they're also used further below?

sjenning replied:
https://github.com/openshift/cloud-credential-operator/pull/185#discussion_r414758727

joelddiaz commented on pkg/operator/staticresource/controller.go:
```diff
@@ -0,0 +1,289 @@
+	}
+	requestedMutatingWebhookConfiguration := ReadMutatingWebhookConfigurationV1Beta1OrDie(v410_00_assets.MustAsset("v4.1.0/aws-pod-identity-webhook/mutatingwebhook.yaml"))
+	requestedMutatingWebhookConfiguration.Webhooks[0].ClientConfig.CABundle = []byte(caBundle)
+	resultMutatingWebhookConfiguration, modified, err := ApplyMutatingWebhookConfiguration(r.clientset.AdmissionregistrationV1beta1(), r.eventRecorder, requestedMutatingWebhookConfiguration, r.mutatingWebhookConfigurationGeneration)
```
What is the purpose of storing and sending the mutatingWebhookConfigurationGeneration?

sjenning replied:
That is the easiest way to detect if something external has changed the `MutatingWebhookConfiguration` as the apiserver will increment the `Generation`.  `ApplyDeployment` does the same thing (rather than try to DeepEquals the `spec` or something like that)

joelddiaz replied:
I suppose my question comes from not seeing ApplyMutatingWebhookConfiguration() do anything with that final parameter that is passed to it.

sjenning replied:
ah yes, very true.  in fact, i'm going to take that out completely in light @marun sharing that i can just add an annotation to the `MutatingWebhookConfiguration` and the `caBundle` will be managed for me.  thus external `Generation` changes would be expected.

dgoodwin replied:
What was this annotation and feature that manages the caBundle for you? I am fighting similar problems over in Hive.

dgoodwin replied:
n/m found it on slack.

sjenning commented:
@marun yes, the service-ca-signer is the CA that signs `Services` that have the `service.beta.openshift.io/serving-cert-secret-name` annotation. It creates a cert/key and injects a secret so that TLS application can be accessed by name on the service network.

sjenning commented on bindata/v4.1.0/aws-pod-identity-webhook/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+        - --port=6443
+        - --service-name=pod-identity-webhook
+        - --tls-secret=pod-identity-webhook
+        - --annotation-prefix=eks.amazonaws.com
```
at some point before release, yes.  could we track this for a follow-on so that we aren't bogged down on naming; the hardest of issues in software development?

joelddiaz replied:
agreed. maybe just stick a # TODO comment if possible.

sjenning commented on bindata/v4.1.0/aws-pod-identity-webhook/mutatingwebhook.yaml:
```diff
@@ -0,0 +1,17 @@
+  name: pod-identity-webhook
+webhooks:
+- name: pod-identity-webhook.amazonaws.com
+  failurePolicy: Ignore
```
This was just copied from upstream.  It is a whole different ball game if you use `Fail` here, as you can take out the cluster by denying pod admission if something goes wrong.

sjenning replied:
I would prefer to keep this as is for now during devel.  I'll track this for revisiting before release.

joelddiaz replied:
Understood about the potentially catastrophic consequences if the webhook starts to error. Just wanted confirmation that I understood why.

sjenning commented on bindata/v4.1.0/aws-pod-identity-webhook/deployment.yaml:
```diff
@@ -0,0 +1,37 @@
+      containers:
+      - name: pod-identity-webhook
+        image: ${IMAGE}
+        imagePullPolicy: Always
```
Yes and yes.

sjenning commented:
/hold cancel

no point since master is locked

sjenning commented:
Updates:
- use `imagePullPolicy: Always`
- use `service.beta.openshift.io/inject-cabundle` to inject service-ca
- create a namespace-local non-manager informer cache the `awspodidentity` controller can do a very narrow watch and we can use `Role` rather than `ClusterRole`
- `awsPodIdentityController` now retries the reconcilation in a loop every 10s before starting the namespace-local informer cache
- replaced `log.WithFields(log.Fields{"controller": controllerName})` with `r.logger`
- removed `Generation` tracking for `MutatingWebhookConfiguration`
- requeue on failed reconcilation
- rename package from `staticresource` to `awspodidentity`

marun commented on pkg/operator/awspodidentity/controller.go:
```diff
@@ -0,0 +1,287 @@
+}
+
+// ApplyMutatingWebhookConfiguration merges objectmeta, does not worry about anything else
+func ApplyMutatingWebhookConfiguration(client admissionregistrationclientv1beta1.MutatingWebhookConfigurationsGetter, recorder events.Recorder, required *admissionregistrationv1beta1.MutatingWebhookConfiguration) (*admissionregistrationv1beta1.MutatingWebhookConfiguration, bool, error) {
```
As written, modification of the webhook config spec (e.g the url) by other actors will not prompt action by the operator. Is this intentional?

sjenning replied:
yes, this is true.  i currently just ensures that the resource exists basically.

since an outside controller injects the service-ca, we really can't use `Generation` as Deployments do to detect changes and external changes are expected.

I would like to check more here.  I'll add a TODO.

sjenning replied:
Actually, I might just add this.  Seems like if you iterate through `wehooks` and clear out `configConfig.caBundle`, we could just do a `DeepEquals` on it.

marun replied:
Please review the documentation for Apply{DaemonSet,Deployment}:

https://github.com/openshift/library-go/blob/master/pkg/operator/resource/resourceapply/apps.go#L21

I think it makes sense to set the input spec hash as an annotation to enable detection of changes across operator releases. 

Regarding the use of generation to detect changes made by other actors, you're right that it can't be used in the same manner as other apply methods due to the CA field being managed by service ca. Since this behavior depends on the service ca injection annotation, it should be possible to detect when to retain the CA field of the existing resource to avoid racing. At worst this implies a second operator-initiated update when the CA is changed by service ca (to pick up the generation). I think that's a small price to pay to ensure that the operator can ensure the intended spec.

cc: @deads2k

sjenning commented:
trying to explain why the `AWS_POD_IDENTITY_WEBHOOK_IMAGE` env var doesn't seem to be set sometimes
```
38m         Normal    ServiceAccountCreated                 deployment/cloud-credential-operator              Created ServiceAccount/pod-identity-webhook -n openshift-cloud-credential-operator because it was missing
38m         Normal    ClusterRoleCreated                    deployment/cloud-credential-operator              Created ClusterRole.rbac.authorization.k8s.io/pod-identity-webhook because it was missing
38m         Normal    RoleCreated                           deployment/cloud-credential-operator              Created Role.rbac.authorization.k8s.io/pod-identity-webhook -n openshift-cloud-credential-operator because it was missing
38m         Normal    ClusterRoleBindingCreated             deployment/cloud-credential-operator              Created ClusterRoleBinding.rbac.authorization.k8s.io/pod-identity-webhook because it was missing
38m         Normal    RoleBindingCreated                    deployment/cloud-credential-operator              Created RoleBinding.rbac.authorization.k8s.io/pod-identity-webhook -n openshift-cloud-credential-operator because it was missing
38m         Normal    ServiceCreated                        deployment/cloud-credential-operator              Created Service/pod-identity-webhook -n openshift-cloud-credential-operator because it was missing
35m         Warning   DeploymentCreateFailed                deployment/cloud-credential-operator              Failed to create Deployment.apps/pod-identity-webhook -n openshift-cloud-credential-operator: Deployment.apps "pod-identity-webhook" is invalid: spec.template.spec.containers[0].image: Required value
30m         Warning   DeploymentCreateFailed                deployment/cloud-credential-operator              Failed to create Deployment.apps/pod-identity-webhook -n openshift-cloud-credential-operator: Deployment.apps "pod-identity-webhook" is invalid: spec.template.spec.containers[0].image: Required value
26m         Warning   DeploymentCreateFailed                deployment/cloud-credential-operator              Failed to create Deployment.apps/pod-identity-webhook -n openshift-cloud-credential-operator: Deployment.apps "pod-identity-webhook" is invalid: spec.template.spec.containers[0].image: Required value
25m         Normal    DeploymentCreated                     deployment/cloud-credential-operator              Created Deployment.apps/pod-identity-webhook -n openshift-cloud-credential-operator because it was missing
25m         Normal    MutatingWebhookConfigurationCreated   deployment/cloud-credential-operator              Created MutatingWebhookConfiguration.admissionregistration.k8s.io/pod-identity-webhook because it was missing

sjenning commented:
Weird, installed a new cluster and now I'm not seeing it.  Added a proper fatal error to capture it though.

sjenning commented:
flake https://bugzilla.redhat.com/show_bug.cgi?id=1817588

/test e2e-aws-upgrade

sjenning commented:
ok, all tests are green.  I've deployed a cluster with the release image from this PR and everything looks good.  I'm going to lock this in and move on to adding the `serviceAccountRef` to the CRD.

joelddiaz commented on pkg/operator/awspodidentity/controller.go:
```diff
@@ -0,0 +1,297 @@
+	err := r.ReconcileResources()
+	if err != nil {
+		r.logger.Errorf("reconciliation failed, retrying in %s", retryInterval.String())
+		return reconcile.Result{RequeueAfter: retryInterval}, err
```
Why a static retry schedule instead of using the built-in exponential back-off?

joelddiaz commented on pkg/operator/awspodidentity/controller.go:
```diff
@@ -0,0 +1,297 @@
+	requestedDeployment := resourceread.ReadDeploymentV1OrDie(v410_00_assets.MustAsset("v4.1.0/aws-pod-identity-webhook/deployment.yaml"))
+	requestedDeployment.Spec.Template.Spec.Containers[0].Image = r.imagePullSpec
+	resultDeployment, modified, err := resourceapply.ApplyDeployment(r.clientset.AppsV1(), r.eventRecorder, requestedDeployment, r.deploymentGeneration, false)
+	r.deploymentGeneration = resultDeployment.Generation
```
Should we read and store the Generation field before checking whether an error ocorred?

sjenning commented:
changes carried on in https://github.com/openshift/cloud-credential-operator/pull/187

And linking back to unaddressed review https://github.com/openshift/cloud-credential-operator/pull/187#issuecomment-623675188

sjenning commented:
reopening as the new head of current development and to get a CI release image with these changes

sjenning commented:
/test e2e-aws

sjenning commented:
these CI release images get reclaimed quickly!

/test e2e-aws
